### PR TITLE
42 check for completeness of a http request and parse accordingly

### LINKED ISSUE #42: Check for completeness of a HTTP request and parse accordingly

### PR DESCRIPTION
**Pull request by u413-284-si**
## Overview
1. Check for complete HTTP request in RequestParser class
2. Introduce enum ParsingStatus
3. Check against method if body detected
4. Check of request size against max body size (determined by ConfigFile) -> behaviour needs to be finished
5. Adapt test cases

### re 1.
- Check for "\r\n\r\n": if not found, abort parsing
- After request line and headers are parsed, determine whether a body needs to be parsed
- If body indicated, check request string for completeness of body -> two cases: body size determined by Content-Length or end of chunked body signalled by zero chunk

### re 2.
- ParsingStatus determines which parsing process shall be performed
- Repeated calls to parser will only trigger indicated parsing process when checks are passed
- Once parsing is completed, the according status triggers the response builder to act

### re 3.
- If a body is indicated by according headers, it is verified that the method can actually have a method. If check failed, throws bad request exception

### re 4.
- Check for too large request sizes and return according error to client
- Added status code 413 in enum and utility functions
- Need to decide whether to close connection in this case -> not specified by nginx

### re 5.
- Test cases needed too be adapted to reflect new check for valid method <-> body combo
- Optimised test cases that expect a throw

closes #42

### CODE REVIEW

File: src/RequestParser.cpp [hunk 1]
```diff
@@ -734,3 +811,91 @@ void RequestParser::checkTransferEncoding(HTTPRequest& request)
...
+ *
+ * @param bodyString The actual body of the HTTP request received so far.
+ * @param request The HTTP request object containing headers and status.
+ *
+ * @return true If the body is complete.
+ * @return false If the body is not yet complete.
+ *
+ * @throws std::runtime_error If the body size exceeds the specified `Content-Length`.
+ */
+bool RequestParser::checkForCompleteBody(const std::string& bodyString, HTTPRequest& request)
```
Comment (QCHR1581) [inline src/RequestParser.cpp:836]: I tried to find out what the behaviour of our server should be, when the content-length is bigger than the body size. 

My search was not unfortunately not successful (looked into RFCs, in the doc of nginx etc.). 

I see that you do not set any specific status code for this case and just normally process such a request. 
Is that on purpose?

**Review by QCHR1581 — CHANGES_REQUESTED** [review]
Great work handling the checking and the parsing of a http request. 

 Concerning functions which return bools:
 
 I would suggest that such function should begin with "is" or "did", like bool vars in general do (according to the webkit style guide).
 
 For example:

- checkForCompleteBody -> isBodyComplete
- checkMethodCanHaveBody -> isMethodAllowedToHaveBody. 

Additionally it would be nice to clarify, which behaviour is correct for the case, that the content-length is bigger than the body.

File: src/RequestParser.cpp (hunk 1 above)
Comment (u413-284-si) [reply src/RequestParser.cpp]: If the received body is less than the indicated Content-Length than this function will never return true, therefore the parsing of the body will never be triggered. Eventually the connection will simply timeout as the server expects more data, but receives no more.

**Comment by u413-284-si** [comment]
## Rework RequestParser
- Remove ParsingStatus enum
- split general parseHttpRequest method into parseHeader() and parseBody()
- add getters for m_hasBody and m_isChunked
- move check for CompleteBody into Server
- fix checks against Content-Length
- fix overwrite of Content-Length in case of a chunked Body

## Rework Server
- adapt connectionReceiveHeader() to new parser methods
- add connectionReceiveBody()
- integrate check for complete Body
- check against clientMaxBodySize -> connectionReceiveBody() differs in this point from  connectionReceiveHeader() which checks against the buffer size
- add status code 413: StatusRequestEntityTooLarge

Furthermore fix unit tests to work with new request parser methods.

File: src/Server.cpp [hunk 2]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+		connection.m_status = Connection::Closed;
+	} else {
+		connection.m_buffer += buffer;
+		if (connection.m_buffer.size() >= Server::s_clientMaxBodySize) {
+			LOG_ERROR << "Maximum allowed client request body size reached from " << connection.m_clientSocket;
+			connection.m_request.httpStatus = StatusRequestEntityTooLarge;
+			connection.m_status = Connection::BuildResponse;
+		}
+		try {
+			if (isCompleteBody(connection.m_buffer, connection.m_request)) {
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/Server.cpp:750]: I just saw that we do expect a throw in isCompleteBody(). 

Still: We can just mimic style of  connectionReceiveHeader(): if false you set HTTPRequest and state of connection. No need for throw.

File: src/Server.cpp [hunk 3]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+{
+	std::map<std::string, std::string>::const_iterator contentLengthIterator = request.headers.find("Content-Length");
+	std::map<std::string, std::string>::const_iterator transferEncodingIterator
+		= request.headers.find("Transfer-Encoding");
+
+	if (contentLengthIterator != request.headers.end() && transferEncodingIterator == request.headers.end()) {
+		unsigned long contentLength = std::strtoul(contentLengthIterator->second.c_str(), NULL, decimalBase);
+		if (contentLength < connectionBuffer.size()) {
+			request.httpStatus = StatusBadRequest;
+			throw std::runtime_error(ERR_CONTENT_LENGTH);
```
Comment (gwolf-011235) [inline src/Server.cpp:789]: This should not throw.

File: test/testBody.cpp [hunk 4]
```diff
@@ -1,96 +1,151 @@
...
-TEST(RequestParser_ValidBody, ChunkedBody) {
-	RequestParser	p;
-	HTTPRequest		request;
+TEST(RequestParser_ValidBody, ChunkedBody)
+{
+	RequestParser p;
+	HTTPRequest request;
 
-	p.parseHttpRequest("GET /search?query=openai&year=2024#conclusion HTTP/1.1\r\nHost: www.example.com\r\nTransfer-Encoding: gzip, chunked\r\n\r\n6\r\nhello \r\n6\r\nworld!\r\n0\r\n\r\n", request);
+	p.parseHeader("POST /search?query=openai&year=2024#conclusion HTTP/1.1\r\nHost: "
```
Comment (gwolf-011235) [inline test/testBody.cpp:11]: I think right now you have to call parseHeader() before parseBody() since state is saved in requestParser. 
If moved into struct no need for that

File: test/testBody.cpp
Comment (gwolf-011235) [inline test/testBody.cpp:1]: General: I suggest to name the test files with the style:
- prefix "test_*" 
- if a class is tested include the class name
- if a class function is tested also include the function name
- if a function is tested include the function name.
E.g. test_RequestParser_parseBody.cpp

You could also create a test fixture TEST_F to get rid of the repeated 
```
RequestParser p;
HTTPRequest request;
```

File: test/testBody.cpp [hunk 5]
```diff
@@ -1,96 +1,151 @@
...
+				p.parseBody("6\r\nhello \n6\r\nworld!\r\n0\r\n\r\n", request);
+			} catch (const std::runtime_error& e) {
+				EXPECT_STREQ("Invalid HTTP request: Method should not have a body", e.what());
+				throw;
+			}
+		},
+		std::runtime_error);
 }
 
 // PREVIOUS TEST CODE
```
Comment (gwolf-011235) [inline test/testBody.cpp:151]: Since deleted from the other files: could also delete here.
The files are tracked with git, so you could just checkout an older version

**Review by gwolf-011235 — CHANGES_REQUESTED** [review]
Nice! Integrates well with existing code.

Two main things: 
1. RequestParser still has state making parsing different connections hard
2. Dont throw if not needed

Something I realised (I'll put into issue): if we want to upload / download a big file (~2GB) we will have a problem since we store everything in an std::string.

File: inc/RequestParser.hpp [hunk 6]
```diff
@@ -23,30 +24,45 @@ class RequestParser {
...
 	RequestParser();
 
-	void parseHttpRequest(const std::string& requestString, HTTPRequest& request);
+	void parseHeader(const std::string& headerString, HTTPRequest& request);
+	void parseBody(const std::string& bodyString, HTTPRequest& request);
 	static void clearRequest(HTTPRequest& request);
 	void clearParser();
 
+	bool hasBody() const;
+	bool isChunked() const;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply inc/RequestParser.hpp]: Changed.

File: inc/RequestParser.hpp [hunk 7]
```diff
@@ -23,30 +24,45 @@ class RequestParser {
...
 	static std::string parseVersion(const std::string& requestLine, HTTPRequest& request);
+
+	// Header Parsing
+	void parseHeaders(HTTPRequest& request);
+
+	// Body Parsing
+	void parseChunkedBody(HTTPRequest& request);
+	void parseNonChunkedBody(HTTPRequest& request);
+
+	// Checks
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply inc/RequestParser.hpp:50]: His comment is regarding functions that return a boolean. Is a certain condition given? Yes/no. Here we have functions which perform more complex checks and which do not return anything. That's why the naming is such.

File: src/RequestParser.cpp [hunk 8]
```diff
@@ -244,33 +252,75 @@ void RequestParser::clearParser()
...
+/**
+ * @brief Parses the header of an HTTP request.
+ *
+ * This function takes a string representation of an HTTP request header.
+ * It extracts the request line and headers from the request string and populates the
+ * provided HTTPRequest object with the parsed data.
+ *
+ * @param headerString The string representation of the HTTP request header.
+ * @param request The HTTPRequest object to populate with the parsed data.
+ */
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:267]: Added.

File: src/RequestParser.cpp [hunk 9]
```diff
@@ -525,12 +587,33 @@ std::string RequestParser::parseVersion(const std::string& requestLine, HTTPRequ
...
+/**
+ * @brief Parses the body of an HTTP request.
+ *
+ * This function is responsible for parsing the body of an HTTP request. It takes a string representation of the body
+ * and populates the provided HTTPRequest object with the parsed data. The parsing logic depends on whether the request
+ * is chunked or non-chunked.
+ *
+ * @param bodyString The string representation of the request body.
+ * @param request The HTTPRequest object to populate with the parsed data.
+ */
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:576]: Added.

File: src/RequestParser.cpp [hunk 10]
```diff
@@ -574,8 +657,11 @@ void RequestParser::parseChunkedBody(HTTPRequest& request)
 			throw std::runtime_error(ERR_CHUNK_SIZE);
 		}
 		request.body += chunkData;
-		// length += numChunkSize;
+		length += numChunkSize;
 	} while (numChunkSize > 0);
+	std::stringstream sstream;
+	sstream << length;
+	request.headers["Content-Length"] = sstream.str();
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:641]: The contrary happens here: the conversion takes place from number to string in order to overwrite the value saved for "Content-Length" in the header map. Therefore the strstream was chosen.

File: src/RequestParser.cpp [hunk 11]
```diff
@@ -729,8 +816,37 @@ void RequestParser::checkTransferEncoding(HTTPRequest& request)
...
+
+	switch (request.method) {
+	case MethodGet:
+	case MethodCount:
+		return false;
+	case MethodPost:
+	case MethodDelete:
+		return true;
+	}
+	return false;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:828]: Added.

File: src/Server.cpp [hunk 12]
```diff
@@ -1,4 +1,8 @@
 #include "Server.hpp"
+#include "Log.hpp"
+#include "RequestParser.hpp"
+#include "StatusCode.hpp"
+#include "error.hpp"
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:5]: Moved.

File: src/Server.cpp [hunk 13]
```diff
@@ -289,21 +293,49 @@ ssize_t Server::writeToSocket(int sockfd, const char* buffer, size_t size, int f
...
+
+/**
+ * @brief Parses the body of an HTTP request.
+ *
+ * This function is responsible for parsing the body of an HTTP request and populating the provided HTTPRequest object
+ * with the parsed data.
+ *
+ * @param bodyString The string representation of the request body.
+ * @param request The HTTPRequest object to be populated with the parsed data.
+ */
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp]: Yep, dunno what happened there, Copilot messed up and I didn't realise it. Changed.

File: src/Server.cpp [hunk 14]
```diff
@@ -628,11 +660,11 @@ void connectionReceiveHeader(Server& server, int clientFd, Connection& connectio
 
 	const ssize_t bytesRead = server.readFromSocket(clientFd, buffer, bytesToRead, 0);
 	if (bytesRead == -1) {
-		// Internal server error
+		LOG_DEBUG << "Internal server error while reading from socket: " << connection.m_clientSocket;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:642]: Changed.

File: src/Server.cpp [hunk 15]
```diff
@@ -628,11 +660,11 @@ void connectionReceiveHeader(Server& server, int clientFd, Connection& connectio
...
 		close(clientFd);
 		connection.m_status = Connection::Closed;
 	} else if (bytesRead == 0) {
-		// Connection closed by client
+		LOG_DEBUG << "Connection closed by client: " << connection.m_clientSocket;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:646]: Changed.

File: src/Server.cpp [hunk 16]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
 {
 	LOG_DEBUG << "ReceiveBody for: " << connection.m_clientSocket;
 
-	(void)server;
-	(void)clientFd;
+	char buffer[Server::s_clientBodyBufferSize] = {};
+
+	const ssize_t bytesRead = server.readFromSocket(clientFd, buffer, sizeof(buffer), 0);
+	if (bytesRead == -1) {
+		LOG_DEBUG << "Internal server error while reading from socket: " << connection.m_clientSocket;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:714]: Changed.

File: src/Server.cpp [hunk 17]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+		close(clientFd);
+		connection.m_status = Connection::Closed;
+	} else if (bytesRead == 0) {
+		LOG_DEBUG << "Connection closed by client: " << connection.m_clientSocket;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:718]: Changed.

File: src/Server.cpp [hunk 18]
```diff
@@ -641,14 +673,19 @@ void connectionReceiveHeader(Server& server, int clientFd, Connection& connectio
 		if (isCompleteRequestHeader(connection.m_buffer)) {
 			LOG_DEBUG << "Received complete request header: " << '\n' << connection.m_buffer;
 			try {
-				server.parseHttpRequest(connection.m_buffer, connection.m_request);
-				server.clearParser();
+				server.parseHeader(connection.m_buffer, connection.m_request);
 			} catch (std::exception& e) {
 				LOG_ERROR << "Error: " << e.what();
+				connection.m_status = Connection::BuildResponse;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:658]: Changed.

File: src/Server.cpp [hunk 19]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+	} else if (bytesRead == 0) {
+		LOG_DEBUG << "Connection closed by client: " << connection.m_clientSocket;
+		close(clientFd);
+		connection.m_status = Connection::Closed;
+	} else {
+		connection.m_buffer += buffer;
+		if (connection.m_buffer.size() >= Server::s_clientMaxBodySize) {
+			LOG_ERROR << "Maximum allowed client request body size reached from " << connection.m_clientSocket;
+			connection.m_request.httpStatus = StatusRequestEntityTooLarge;
+			connection.m_status = Connection::BuildResponse;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp:726]: Good point. Added just as in connectionReceiveHeader().

File: src/Server.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp]: Implemented.

File: src/Server.cpp [hunk 20]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+				server.parseBody(connection.m_buffer, connection.m_request);
+				connection.m_status = Connection::BuildResponse;
+				server.modifyEvent(clientFd, EPOLLOUT);
+				connection.m_buffer.erase(0);
+			} else
+				LOG_DEBUG << "Received partial request body: " << '\n' << connection.m_buffer;
+		} catch (std::exception& e) {
+			LOG_ERROR << "Error: " << e.what();
+			connection.m_status = Connection::BuildResponse;
+			return;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp]: Changed, non regarding if the body parsing was successful or an error was thrown, we need to perform the same three steps:

- connection.m_status = Connection::BuildResponse;
- server.modifyEvent(clientFd, EPOLLOUT);
- connection.m_buffer.erase(0);

Timestamp is set earlier after "connection.m_buffer += buffer;". No extra return needed as the function terminates anyway after that.

File: src/Server.cpp (hunk 3 above)
Comment (u413-284-si) [reply src/Server.cpp]: Removed.

File: src/Server.cpp [hunk 21]
```diff
@@ -673,12 +710,88 @@ bool isCompleteRequestHeader(const std::string& connectionBuffer)
...
+ * A complete request body is determined by the presence of the "Content-Length" header or the "Transfer-Encoding"
+ * header. If the "Content-Length" header is present, the function checks if the size of the connection buffer matches
+ * the specified content length. If the "Transfer-Encoding" header is present, the function checks if the connection
+ * buffer contains the end marker "0\r\n\r\n".
+ *
+ * @param connectionBuffer The buffer to check for a complete request body.
+ * @param request The HTTPRequest object containing the request headers.
+ * @return true if the buffer contains a complete request body, false otherwise.
+ */
+bool isCompleteBody(const std::string& connectionBuffer, HTTPRequest& request)
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp]: These are two different things:

In the RequestParser we establish whether or not the request has a body and whether or not it is chunked. We save it as a state description in the HTTPRequest.

Then we need to determine whether the received body is complete or not. And here we need to take into consideration different requirements for each case: chunked or non-chunked. That is performed with this function. What you describe is the next step once we have ensured that the body is complete. We call the appropriate method to parse the body. This is done in parseBody():

```
void RequestParser::parseBody(const std::string& bodyString, HTTPRequest& request)
{
	m_requestStream.str(bodyString);
	if (request.isChunked)
		parseChunkedBody(request);
	else
		parseNonChunkedBody(request);
	resetRequestStream();
}
``

File: test/testBody.cpp (hunk 4 above)
Comment (u413-284-si) [reply test/testBody.cpp:11]: Well, I do it because I want to determine whether the body is chunked or not. And also to retrieve the content-length. 

I could also just set it in the HTTPRequest and not parse the header. If need be, I can change.

File: test/testBody.cpp (hunk 5 above)
Comment (u413-284-si) [reply test/testBody.cpp:151]: Yeah, forgot to delete here. Removed.

File: inc/RequestParser.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply inc/RequestParser.hpp:54]: True, let me rephrase: I think "check" implies a simple check, which would with Chris' proposal have a is/has name. Therefore to prefix a function "check" isn't possible.

I would go for a different prefix like _validate_

Right now the check functions combine a check and state change (throw as convenience). They don't need to return because they throw. I think complexity of a check doesn't change the nature of it per se. It's just harder to answer the question e.g. is this a valid header? Even if I don't think most of them perform complex checks. checkContentLength is a bit more complex, since it also does the multiple header shebang.

I think cleaner would be to have the check function return a bool (since it checks). The caller would then decide what to do with the information (most likely state change and throw if false).
This would loose convenience, so easier would be to not use prefix check

File: src/RequestParser.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp:635]: Yes I misread! 
In utilities there is a templated toString() function which could be used instead. Though I don't like it because it just seems heavy to me.

File: src/Server.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/Server.cpp]: Why do we need to check for the headers Content-Length and Transfer-Encoding
```
std::map<std::string, std::string>::const_iterator contentLengthIterator
		= connection.m_request.headers.find("Content-Length");
	std::map<std::string, std::string>::const_iterator transferEncodingIterator
		= connection.m_request.headers.find("Transfer-Encoding");
```
If I understood correctly the parser makes sure that the request headers are correct. It also sets 
- hasBody (meaning we have either Content-Length or Transder-Encoding) and 
- isChunked (meaning we have chunked encoding).

So the condition `If (HTTPRequest.hasBody && !HTTPRequest.isChunked)` checks for the case that Content-Length is given and transfer is not chunked (as already determined by the parser). We only need to extract out content length 
`unsigned long contentLength = std::strtoul(connection.m_request.headers.at("Content-Length"), NULL, decimalBase);`

If this condition doesn't fire we know that we have chunked encoding (as the parser has checked). So we can just return `connection.m_buffer.find("0\r\n\r\n") != std::string::npos;`

So we have 
```
If (HTTPRequest.hasBody && !HTTPRequest.isChunked) {
  unsigned long contentLength = std::strtoul(connection.m_request.headers.at("Content-Length"), NULL, decimalBase);
  if (contentLength == connection.m_buffer.size())
    return true;
  if (contentLength < connection.m_buffer.size())
    connection.m_request.httpStatus = StatusBadRequest;
  return false;
} else {
 return connection.m_buffer.find("0\r\n\r\n") != std::string::npos;
} 
```

We could also save the Content-Length in the HTTP Request in the parser, so that we don't have to convert everytime.

File: inc/RequestParser.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply inc/RequestParser.hpp:54]: Renamed with prefix validate.

File: src/RequestParser.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:635]: Good hint, forgot about this one, replaced.

File: src/Server.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/Server.cpp]: I see what you mean. Thought you wanted to jump this step completely and go to the parsing directly.

Yeah, we can just use the booleans instead of searching for the map entries. We can even omit the first "HTTPRequest.hasBody" check as we already know in this function call that there is a body. Reduced the function to simple var checks.

File: test/testBody.cpp (hunk 4 above)
Comment (u413-284-si) [reply test/testBody.cpp]: Test are only calling parseBody() now. The required conditions are set in the test fixture.

File: test/testBody.cpp
Comment (u413-284-si) [reply test/testBody.cpp:1]: Renamed test files. Introduced test fixture in all tests. Introduced Arrange, Act, Assert (AAA) pattern for unit testing.

**Comment by u413-284-si** [comment]
Had to move timestamp to function end again to account for the other cases when readFromSocket returns. Added timestamp to error cases just before the return statement.

**Review by gwolf-011235 — APPROVED** [review]

File: src/RequestParser.cpp (hunk 1 above)
Comment (QCHR1581) [reply src/RequestParser.cpp]: Thanks for the clarification. Makes senses.

**Review by QCHR1581 — APPROVED** [review]
